torch/_inductor/autotune_process.py

```diff
@@ -131,15 +127,121 @@ def close(self) -> None:
 
     def terminate(self) -> None:
         """
-        Signal the child process to terminate and wait for it to exit.
+        Signal the child process to terminate.
         """
         if self.process is not None:
             self.put(None)
+
+    def wait(self) -> None:
+        """
+        Wait for the child process to exit.
+        """
+        if self.process is not None:
             self.process.wait()
             self.close()
 
 
-tuning_process = TuningProcess()
+@dataclasses.dataclass
+class TuningProcessPool:
+    processes: Optional[Queue[TuningProcess]] = None
+    executor: Optional[ThreadPoolExecutor] = None
+
+    def initialize(self, count: Optional[int] = None) -> None:
+        """
+        Start the child processes.
+        """
+        assert (self.processes is None) == (self.executor is None)
+        if self.processes is not None:
+            return
+
+        count = count or torch.cuda.device_count()
+        assert count > 0
+
+        # Launch the child processes and push a msg to "warm up"
+        self.processes = Queue()
+        for device in range(count):
+            p = TuningProcess(device=device)
+            p.initialize()
+            p.put(Ping())
+            self.processes.put(p)
+
+        # Wait for the initialization to finish
+        for p in self.processes.queue:
+            assert isinstance(p.get(), Pong)
+
+        # Use a thread pool to manage distributing work to the subprocesses.
+        # Threads block on an available process, so it makes sense to match
+        # the number of threads with the number of devices.
+        self.executor = ThreadPoolExecutor(max_workers=count)
+
+        # Register the exit handler for the parent process so it will terminate
+        # the child processes.
+        global EXIT_HANDLER_REGISTERED
+        if not EXIT_HANDLER_REGISTERED:
+            EXIT_HANDLER_REGISTERED = True
+            import atexit
```

Comment:
this can be a top level import right

Reply:
@eellison do you just mean to move it out of the class or do I misunderstand? I stuck with the pattern already established. I assume it was done this way to avoid the atexit registration if we never actually initialize and use the sub-process path?

Reply:
Yea, nevermind.